### PR TITLE
Handle OSError exceptions when trying to remove recovery checkpoints

### LINKED ISSUE #32065: WinError 145 on closing Mantid
**Describe the bug**
On closing Mantid the following error is sometimes occuring

```
 File "C:\MantidNightlyInstall\bin\workbench\app\mainwindow.py", line 588, in closeEvent
   self.project_recovery.remove_current_pid_folder()
 File "C:\MantidNightlyInstall\bin\workbench\projectrecovery\projectrecovery.py", line 199, in remove_current_pid_folder
   self._remove_directory_and_directory_trees(self.recovery_directory_pid)
 File "C:\MantidNightlyInstall\bin\workbench\projectrecovery\projectrecovery.py", line 138, in _remove_directory_and_directory_trees
   shutil.rmtree(directory)
 File "C:\MantidNightlyInstall\bin\lib\shutil.py", line 730, in rmtree
   return _rmtree_unsafe(path, onerror)
 File "C:\MantidNightlyInstall\bin\lib\shutil.py", line 603, in _rmtree_unsafe
   _rmtree_unsafe(fullname, onerror)
 File "C:\MantidNightlyInstall\bin\lib\shutil.py", line 612, in _rmtree_unsafe
   onerror(os.rmdir, path, sys.exc_info())
 File "C:\MantidNightlyInstall\bin\lib\shutil.py", line 610, in _rmtree_unsafe
   os.rmdir(path)
OSError: [WinError 145] The directory is not empty: '...\mantidproject\\mantid\\workbench-recovery\....'
```

**To Reproduce**
The issue occurs when closing Mantid but not everytime and is therefore difficult to reproduce.

**Expected behavior**
When closing Mantid there should be no crashes or errors.

**Platform/Version (please complete the following information):**
Mantid version 5.0.20200909.1539 
Mantid version 6.1.20210719.2059

Windows

### PR DESCRIPTION
**Pull request by martyngigg**
**Description of work.**

We now trap errors that would have caused the error reporter to be raised. A user can do nothing about
the errors and recovery cleans up after itself in future runs so the risk is minimal. See the commit messages
for more details.

**To test:**

* Please pay close attention to code changes amd review them throughly to check the logic.
* Run through the [project recovery manual tests](https://developer.mantidproject.org/Testing/ErrorReporter-ProjectRecovery/ProjectRecoveryTesting.html)

Fixes #32065 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Review by DanielMurphy22 — CHANGES_REQUESTED** [review]
Just one small code comment I want to check. Manual testing of Project Recovery worked fine.

File: qt/applications/workbench/workbench/projectrecovery/projectrecovery.py [hunk 1]
```diff
@@ -107,39 +107,34 @@ def stop_recovery_thread(self):
...
-                if len(os.listdir(folder)) == 0:
-                    self._remove_directory_and_directory_trees(folder)
-            except OSError:
-                # Fail silently as expected for all folders
-                pass
+            if len(os.listdir(folder)) == 0:
+                self._remove_directory_and_directory_trees(folder, ignore_errors=True)
 
-    def _remove_directory_and_directory_trees(self, directory):
+    def _remove_directory_and_directory_trees(self, directory: str, *, ignore_errors: bool = False):
```
[the remark this answers is not in the document]
Comment (martyngigg) [reply qt/applications/workbench/workbench/projectrecovery/projectrecovery.py:120]: Yeah it's a nice feature that I especially like for flag-like options like this were just putting `True` is potential confusing.

File: qt/applications/workbench/workbench/app/mainwindow.py [hunk 2]
```diff
@@ -612,6 +611,15 @@ def closeEvent(self, event):
...
                 self.workspacecalculator.view.closeEvent(event)
 
+            if self.project_recovery is not None:
+                # Do not merge this block with the above block that
+                # starts with the same check
+                # We deliberately split the call to stop the recovery
+                # thread and removal of the checkpoints folder to
+                # allow for the maximum amount of time for the recovery
+                # thread to finish
+                self.project_recovery.remove_current_pid_folder()
```
[the remark this answers is not in the document]
Comment (martyngigg) [reply qt/applications/workbench/workbench/app/mainwindow.py:621]: Ah yes thank you! That's the whole error in the first place! 🤦🏻

[the remark this answers is not in the document]
Comment (martyngigg) [reply qt/applications/workbench/workbench/app/mainwindow.py]: That should be fixed now.

[the remark this answers is not in the document]
Comment (martyngigg) [reply qt/applications/workbench/workbench/app/mainwindow.py]: I fixed it and then my push failed. It should be good now.

**Review by DanielMurphy22 — APPROVED** [review]
Looks good to me! Hopefully this will prevent the problem users have been facing!

**Review by DavidFair — APPROVED** [review]
:crossed_fingers: hopefully this resolves this for good

**Pull request merged**